### PR TITLE
Migrate attach max synthetic during and after VM provision.

### PR DESCRIPTION
**Pull request by LiliDeng**
See commit message for details.

### CODE REVIEW

File: lisa/nic.py [hunk 1]
```diff
@@ -143,7 +145,7 @@ def get_unpaired_devices(self) -> List[str]:
         return [x.upper for x in self.nics.values() if not x.lower]
 
     def get_upper_nics(self) -> List[str]:
-        return [x.upper for x in self.nics.values() if x.lower]
+        return [x.upper for x in self.nics.values()]
```
Comment (squirrelsc) [inline lisa/nic.py:148]: Please check with Matthew, if it's a breaking change.

File: lisa/nic.py [hunk 2]
```diff
@@ -119,6 +119,8 @@ class Nics(InitializableMixin):
         )
     )
 
+    __file_not_exist = re.compile(r".*No such file or directory", re.MULTILINE)
```
Comment (squirrelsc) [inline lisa/nic.py]: This class is not a super class of others. So one underscore is enough to mark private variables.

Comment (squirrelsc) [reply lisa/nic.py]: The leading ".*" should be able to remove, it doesn't affect in this pattern.

File: lisa/nic.py [hunk 3]
```diff
@@ -235,22 +237,23 @@ def _get_node_nic_info(self) -> None:
...
-                raise LisaException(
-                    "Could not locate default network interface"
-                    f" in output:\n{result.stdout}"
-                )
-            upper_nic, lower_nic, pci_slot = match.groups()
-            nic_info = NicInfo(upper_nic, lower_nic, pci_slot)
-            self.append(nic_info)
+        result = self._node.execute(nic_info_fetch_cmd, shell=True)
+        if self.__file_not_exist.match(result.stdout):
+            self._node.log.debug(f"There is no VF on node {self._node.name}.")
```
Comment (squirrelsc) [inline lisa/nic.py]: Do we need this log line? It looks not provide many information.

File: lisa/nic.py [hunk 4]
```diff
@@ -235,22 +237,23 @@ def _get_node_nic_info(self) -> None:
         # use sysfs to gather upper/lower nic pairings and pci slot info
         nic_info_fetch_cmd = "ls -la /sys/class/net/*/lower*/device"
         self._node.log.info(f"Gathering NIC information on {self._node.name}.")
```
Comment (squirrelsc) [inline lisa/nic.py]: Change this line to debug level.

File: lisa/nic.py (hunk 1 above)
Comment (sharsonia) [reply lisa/nic.py:148]: no this shouldn't be a breaking change. We need to remove this coupling if we want to use this class as generic class for both Synthetic and Sriov nic.
Upper nic can be the synthetic nic and might not be paired with a lower nic if AN is not enabled.

File: lisa/nic.py (hunk 2 above)
Comment (LiliDeng) [reply lisa/nic.py]: update for both comments

File: lisa/nic.py (hunk 3 above)
Comment (LiliDeng) [reply lisa/nic.py]: removed

File: lisa/nic.py (hunk 4 above)
Comment (LiliDeng) [reply lisa/nic.py]: changed it into debug level

**Review by squirrelsc — APPROVED** [review]

**Pull request merged**